elements/pl-dataframe/pl-dataframe.py

```diff
@@ -0,0 +1,166 @@
+    varname = pl.get_string_attrib(element, "params-name")
+    show_index = pl.get_boolean_attrib(element, "show-index", SHOW_INDEX_DEFAULT)
+    show_header = pl.get_boolean_attrib(element, "show-header", SHOW_HEADER_DEFAULT)
+    show_python = pl.get_boolean_attrib(element, "show-python", SHOW_PYTHON_DEFAULT)
+    show_dimensions = pl.get_boolean_attrib(
+        element, "show-dimensions", SHOW_DIMENSIONS_DEFAULT
+    )
+    show_dtype = pl.get_boolean_attrib(element, "show-dtype", SHOW_DTYPE_DEFAULT)
+
+    display_language = pl.get_enum_attrib(
+        element, "display-language", DisplayLanguage, DISPLAY_LANGUAGE_DEFAULT
+    )
+
+    num_digits = pl.get_integer_attrib(element, "digits", NUM_DIGITS_DEFAULT)
+
+    if varname not in data["params"]:
+        raise KeyError(
+            f'Could not find {varname} in params. Please make sure to set params-name="{varname}" in the element.'
+        )
+
+    # Always assume that entry in params dict is serialized dataframe
+    frame = pl.from_json(data["params"][varname])
+
+    if not isinstance(frame, pd.DataFrame):
+        raise ValueError(f"Parameter name '{varname}' does not encode a dataframe.")
+
+    frame = cast(pd.DataFrame, frame)
+
+    frame_style = frame.style
+
+    # Format integers using commas every 3 digits
+    integer_column_names = frame.select_dtypes(include="int").columns
+    frame_style.format(subset=integer_column_names, thousands=",")
+
+    if num_digits is not None:
+        # Get headers for all floating point columns and style them to use the desired number of sig figs.
+        float_column_names = frame.select_dtypes(include="float").columns
+
+        # This format string displays the desired number of digits, as given by the instructor
+        # Switches between exponential and decimal notation as needed
+        frame_style.format(subset=float_column_names, formatter=f"{{:.{num_digits}g}}")
+
+    if show_dtype:
+        if display_language is DisplayLanguage.PYTHON:
+            get_dtype_function = get_pandas_dtype
+        elif display_language is DisplayLanguage.R:
+            get_dtype_function = convert_pandas_dtype_to_r
+        else:
+            assert_never(show_dtype)
+
+        # old agg function:
+        descriptors = frame.agg([get_dtype_function]).set_axis(
+            ["dtype"], axis="index", copy=False
+        )
+        other = descriptors.style.applymap(lambda v: "font-weight: bold;")
+        frame_style.set_table_styles(
+            [{"selector": ".foot_row0", "props": "border-top: 1px solid black;"}]
+        )
+        frame_style.concat(other)
+
+    if not show_header:
+        frame_style.hide(axis="columns")
+
+    if show_index:
+        # Switch row indices to being 1-based if supplied indices are integer. Otherwise, leave untouched.
+        # TODO double check if this will cause unexpected behavior if these aren't just indices.
+        if display_language is DisplayLanguage.R and pd.api.types.is_integer_dtype(
```

Comment:
@coatless I had to change your code slightly for this to work (not sure why the styler didn't like having its index set directly, kinda weird). But I was wondering, won't this cause issues for non-default indexing? Should we instead check for the index being the numbers 0 to n?

Reply:
@eliotwrobson thanks for the question. I can understand why the linter wouldn't like the shift to 1-based indices. The other way around it would be to switch away from integer indices to string indices: 

```python
frame_style.index = [str(idx) for idx in range(1, len(frame_style.index) + 1)]
```

With that being said, the index shift is 100% cosmetic. Why it's cosmetic is we're not shifting indices and, then, giving the data frame to another element. Everything is happening in a localized and contained manner. Moreover, we can _rename_ and change indices without worrying about subsetting (under the `.loc` case as we're dealing with labels, the `.iloc` case would still be 0-based as it is positional).

```python
import pandas as pd 

## Instantiate a dataframe 
df = pd.DataFrame(
    [[1, "tall"], [2, "small"], [3, "normal"]],
    columns=["id", "height"]
)

print(df)
#    id  height
# 0   1    tall
# 1   2   small
# 2   3  normal

df.loc[1, :]
# id            2
# height    small
# Name: 1, dtype: object

## Create a copy of the dataframe
df_shift_index = df.copy()

## Shift indices
df_shift_index.index = list(range(1, df.shape[0] + 1))
print(df_shift_index)
#   id  height
# 1   1    tall
# 2   2   small
# 3   3  normal

df_shift_index.loc[1,:]
# id           1
# height    tall
# Name: 1, dtype: object
```

So, no issues should arise as the data is localized to the element itself.

Reply:
Right, I understand that this code here is making a cosmetic change, the issue I was wondering about is whether, if using a dataframe that has an integer index which is not just numbering the rows from 0 to n-1, we still want to add 1 to all the values. I think we can make a more robust check on what the index values are if necessary though.

Reply:
Gotcha, how about adding: 

```python
type(frame_style.index) == pd.RangeIndex
```

where `frame_style.index` should have type `pandas.core.indexes.range.RangeIndex`.

If an index is supplied, then `type(frame_style.index)` should have:

```
pandas.core.indexes.base.Index
```

Reply:
Unfortunately, the re-serialization erases the RangeIndex type and recodes it as an Int64. Maybe we can add a note about this to make sure nobody using R runs into this problem?

Reply:
@eliotwrobson how about the following to detect non-standard indices:

```python
def is_default_index_type(df):
  return df.index.is_integer()

def is_default_numeric_index_values(df):
  return pd.Index(list(range(0, len(df)))).equals(df.index)

def is_default_index(df):
  return is_default_index_type(df) and is_default_numeric_index_values(df)

# Sample call
is_default_index(df_zero_based)
```

Experiments:

https://colab.research.google.com/github/coatless/colab-notes/blob/main/04-pandas-index-default-checks.ipynb#scrollTo=5eQr9mCSWwBu

Reply:
Added this in and seems to work correctly. Thanks!